### PR TITLE
feat(client):  add JSON filtering to Postgres and MySQL

### PR DESCRIPTION
**Pull request by williamluke4**

### CODE REVIEW

**Comment by Jolg42** [comment]
Merge of engine will be done by renovate here (if tests pass): https://github.com/prisma/prisma/pull/7118

Would be nice to add a test for this feature in client though.

File: src/packages/client/src/__tests__/integration/happy/exhaustive-schema/__snapshots__/test.ts.snap [hunk 1]
```diff
@@ -16092,14 +16121,48 @@ export namespace Prisma {
...
   }
+  export type JsonNullableFilter = 
+    | PatchUndefined<
+        Either<Required<JsonNullableFilterBase>, Exclude<keyof Required<JsonNullableFilterBase>, 'path'>>,
+        Required<JsonNullableFilterBase>
+      >
+    | OptionalFlat<Omit<Required<JsonNullableFilterBase>, 'path'>>
 
-  export type JsonNullableFilter = {
+  export type JsonNullableFilterBase = {
```
Comment (Jolg42) [inline src/packages/client/src/__tests__/integration/happy/exhaustive-schema/__snapshots__/test.ts.snap:16153]: Maybe something we can mention to typegraphql

File: src/packages/client/src/__tests__/integration/happy/json-filtering-postgres/test.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            number: 1,
+            array: [1, 2, 3, 4],
+          },
+        },
+        {
+          email: 'b@b.de',
+          json: {
+            string: 'bc',
+            number: 2,
+            array: [5, 6, 7, 8],
```
Comment (Jolg42) [inline src/packages/client/src/__tests__/integration/happy/json-filtering-postgres/test.ts:33]: Does it work with an object?

File: src/packages/client/src/generation/TSClient/TSClient.ts [hunk 3]
```diff
@@ -262,7 +262,27 @@ ${prismaEnums.join('\n\n')}
  */
 
 ${this.dmmf.inputObjectTypes.prisma
-  .map((inputType) => new InputType(inputType, collector).toTS())
+  .reduce((acc, inputType) => {
```
Comment (Jolg42) [inline src/packages/client/src/generation/TSClient/TSClient.ts:265]: Looks like a lot, maybe could use a comment so it's easier to read/understand the code?

**Review by Jolg42 — APPROVED** [review]
Nice! 🎊

**Comment by Jolg42** [comment]
Looks like prettier disapproves something 😄 
```
[2021-05-18T13:58:31Z] client /app/src/packages/client/src/__tests__/integration/happy/json-filtering-mysql/test.ts
[2021-05-18T13:58:31Z] client   25:22  error  Replace `a:·1,·b:·2,·c:·3}` with `·a:·1,·b:·2,·c:·3·},`  prettier/prettier
[2021-05-18T13:58:31Z] client   34:22  error  Replace `a:·4,·b:·5,·c:·6}` with `·a:·4,·b:·5,·c:·6·},`  prettier/prettier
[2021-05-18T13:58:31Z] client   43:22  error  Replace `a:·7,·b:·8,·c:·9}` with `·a:·7,·b:·8,·c:·9·},`  prettier/prettier
[2021-05-18T13:58:31Z] client   67:1   error  Delete `····`                                            prettier/prettier
[2021-05-18T13:58:31Z] client   68:21  error  Delete `·`                                               prettier/prettier
[2021-05-18T13:58:31Z] client   80:38  error  Delete `⏎`                                               prettier/prettier
```

**Comment by janpio** [comment]
The now failing tests look pretty much unrelated. Weird. Maybe this is currently actually broken on `master`. Started a CI run of latest commit here: https://github.com/prisma/prisma/runs/2613889751

**Comment by Jolg42** [comment]
Related to https://github.com/prisma/prisma/issues/7135

**Pull request merged**